### PR TITLE
feat: add doppelganger model

### PR DESCRIPTION
**Pull request by ricardodcpereira**
This PR adds the DoppelGANger model to ydata-synthetic. Please find attached a usage example.

[DoppelGANger.pdf](https://github.com/ydataai/ydata-synthetic/files/12242166/DoppelGANger.pdf)

### CODE REVIEW

File: src/ydata_synthetic/synthesizers/timeseries/doppelganger/doppelganger.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+        for i in range(self.num_packing):
+            (g_output_feature_train_tf, g_output_attribute_train_tf,
+             g_output_gen_flag_train_tf, g_output_length_train_tf,
+             g_output_argmax_train_tf) = \
+                self.generator.build(
+                    self.g_real_attribute_input_noise_train_pl_l[i],
+                    self.g_addi_attribute_input_noise_train_pl_l[i],
+                    self.g_feature_input_noise_train_pl_l[i],
+                    self.g_feature_input_data_train_pl_l[i])
+            # g_output_feature_train_tf: batch_size * (time * sample_len) * dim
```
Comment (aquemy) [inline src/ydata_synthetic/synthesizers/timeseries/doppelganger/doppelganger.py:150]: needed?

File: src/ydata_synthetic/synthesizers/timeseries/doppelganger/network.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
+import tensorflow as tf
+import numpy as np
+
+# Adapted from https://github.com/fjxmlzn/DoppelGANger/blob/master/gan/network.py.
```
Comment (aquemy) [inline src/ydata_synthetic/synthesizers/timeseries/doppelganger/network.py:4]: Same, either removed or in a doctstring. Base level inline comments will always trigger warnings by linter and other tools.

**Review by aquemy — CHANGES_REQUESTED** [review]
How can I save and load the model?

Other than that, it looks good - just very minor remarks

File: src/ydata_synthetic/synthesizers/timeseries/doppelganger/doppelganger.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
+import tensorflow as tf
+import numpy as np
+from tqdm import tqdm
+import math
+
+# Adapted from https://github.com/fjxmlzn/DoppelGANger/blob/master/gan/doppelganger.py.
```
[the remark this answers is not in the document]
Comment (ricardodcpereira) [reply src/ydata_synthetic/synthesizers/timeseries/doppelganger/doppelganger.py]: Done.

File: src/ydata_synthetic/synthesizers/timeseries/doppelganger/doppelganger.py (hunk 1 above)
Comment (ricardodcpereira) [reply src/ydata_synthetic/synthesizers/timeseries/doppelganger/doppelganger.py]: Done.

Comment (ricardodcpereira) [reply src/ydata_synthetic/synthesizers/timeseries/doppelganger/doppelganger.py]: Removed.

File: src/ydata_synthetic/preprocessing/timeseries/doppelganger_processor.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
+"Implementation of a DoppelGANger Data Processor."
```
[the remark this answers is not in the document]
Comment (ricardodcpereira) [reply src/ydata_synthetic/preprocessing/timeseries/doppelganger_processor.py]: Removed.

File: src/ydata_synthetic/synthesizers/timeseries/doppelganger/network.py (hunk 2 above)
Comment (ricardodcpereira) [reply src/ydata_synthetic/synthesizers/timeseries/doppelganger/network.py]: Done.

File: src/ydata_synthetic/synthesizers/timeseries/model.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    @property
+    def function(self):
+        return self.__MAPPING__[self.value]
+
+class TimeSeriesSynthesizer():
+    "Abstraction class "
+    def __new__(cls, modelname: str, model_parameters=None, **kwargs):
+        return Model(modelname).function(model_parameters, **kwargs)
```
[the remark this answers is not in the document]
Comment (ricardodcpereira) [reply src/ydata_synthetic/synthesizers/timeseries/model.py:31]: Both the save and load functions were added.

**Review by aquemy — APPROVED** [review]

**Pull request merged**